functions/functions.go

```diff
@@ -0,0 +1,187 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package functions provides objective functions for testing optimization
+// algorithms.
+//
+// We encourage outside contributions of additional test functions that exhibit
+// properties not already covered in the testing suite or with significance due
+// to use in benchmark cases.
+package functions
+
+import (
+	"math"
+	"testing"
+
+	"github.com/gonum/diff/fd"
+	"github.com/gonum/floats"
+)
+
+// Function represents an objective function.
+type Function interface {
+	F(x []float64) float64
+}
+
+type Gradient interface {
+	Df(x, grad []float64)
+}
+
+type FunctionGradient interface {
+	FDf(x, grad []float64) float64
+}
+
+// Minimumer is an objective function that can also provide information about
+// its minima.
+type Minimumer interface {
+	Function
+
+	// Minima returns _known_ minima of the function.
+	Minima() []Minimum
+}
+
+// Minimum represents information about an optimal location of a function.
+type Minimum struct {
+	// X is the location of the minimum. X may not be nil.
+	X []float64
+	// F is the value of the objective function at X.
+	F float64
+	// Global indicates whether the minimum is local or global.
+	Global bool
+}
+
+type FuncTest struct {
+	X []float64
+
+	WantF    float64
+	WantGrad []float64 // If nil, gradient is not evaluated.
+}
+
+// TODO: Decide and implement an exported testing function:
+// func Test(f Function, ??? ) ??? {
+// }
+
+const (
+	defaultTol     = 1e-8
+	defaultGradTol = 1e-5
```

Comment:
There should be two gradient tolerances. One for the specifed gradient value, which is tight (1e-14 ish), and one for finite difference which is loose (1e-6).

I think the function tolerance should be increased to 1e-14. We're comparing to a supposedly known value, so we just have to prevent against floating point noise.